File: builds/01-invoice-extractor/ts/src/pdfText.ts
/** Extract text from a PDF. Digital PDFs only; scanned PDFs need OCR (out of scope for v1, see README). */
import { readFile } from "node:fs/promises";
import { PDFParse } from "pdf-parse";

export async function pdfToText(path: string): Promise<string> {
  const data = await readFile(path);
  const parser = new PDFParse({ data });
  try {
    const result = await parser.getText();
    const text = result.text ?? "";
    if (text.trim().length < 40) {
      throw new Error(`${path}: almost no text layer; likely a scanned PDF (needs OCR)`);
    }
    return text;
  } finally {
    await parser.destroy();
  }
}

File: builds/01-invoice-extractor/extractor/pdf_text.py
"""Extract text from a PDF. Digital PDFs only; scanned PDFs need OCR (out of scope for v1, see README)."""

from pathlib import Path

from pypdf import PdfReader


def pdf_to_text(path: str | Path) -> str:
    reader = PdfReader(str(path))
    text = "\n".join((page.extract_text() or "") for page in reader.pages)
    if len(text.strip()) < 40:
        raise ValueError(f"{path}: almost no text layer; likely a scanned PDF (needs OCR)")
    return text
